### PR TITLE
initial implemention of verifyWith

### LINKED ISSUE #60: (placeholder) Add support for `verifyWith`
`verifyWith` is used by managed delivery to run one-off jobs primarily to run tests and do validations. For the k8s plugin, we need to enable the support for it in such a way that one-off jobs can run as Kubernetes jobs.

### PR DESCRIPTION
**Pull request by nabuskey**
initial implementation of verify with kubernetes job
It uses Spinnaker's runJob (runJobManifest) stage to run a kubernetes job. If the stage fails, it's considered a failure of verification.
Manifest name uses the [`generateName`](https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values) field and the name field supplied is ignored.
If the manifest is supplied with `generateName` field, its value is left alone.
Job name is generated as `verify-<APPLICATION_NAME>-<ENVIRONMENT_NAME>-<ARTIFACT_VERSION>`. 
Users can pass manifest directly with the `manifest` field. 
The delivery config yaml file will looks something like this:
```yaml
name: test1
application: ledemo
serviceAccount: keeltest-service-account
artifacts:
- name: nabuskey/led-demo
  type: docker
  reference: my-docker-artifact
  tagVersionStrategy: increasing-tag
environments:
  - name: dev
    verifyWith: 
    - account: deploy-experiments
      jobNamePrefix: my-job-name # Optional. If supplied, job names created by the plugin will start with this. 
      type: k8s/job@v1
      manifest: 
        metadata:
          name: not-used # the name field is removed
        spec:
          template:
            spec:
              containers:
              - name: pi
                image: alpine
                command: ["ash",  "-c", "sleep 5; exit 0 "]
              restartPolicy: Never
          backoffLimit: 2
    locations:
      account: deploy-experiments
      regions: []
    resources:
    - kind: k8s/resource@v1
      spec:
        container:
          reference: my-docker-artifact
        metadata: 
          application: ledemo
        template:
          apiVersion: apps/v1
          kind: Deployment
          metadata: 
            name: ledemo
            namespace: test
          spec:
            replicas: 1
            selector:
              matchLabels:
                app: ledemo
            template:
              metadata:
                labels:
                  app: ledemo
              spec:
                containers:
                - name: ledemo
                  image: my-docker-artifact

```
Fixes #60

### CODE REVIEW

**Comment by nabuskey** [comment]
Removed `apiVersion` and `kind` requirements from manifest.
Add optional `jobNamePrefix` field.

**Comment by dmart** [comment]
The generated job name is likely to be longer than 63 (as in our testing) characters which causes the job to fail.

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (dmart) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:122]: This can easily result in a job name being > 63 characters which causes the job to fail.

Comment (nabuskey) [reply k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:122]: Ah good point. I thought the limit was 263 but I was wrong.

**Comment by JonGilmore** [comment]
fwiw, this is working for us, tested with ec2 debian deploys running verifications as k8s jobs.

**Review by dmart — APPROVED** [review]

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:108]: let's add comments on how this `metadata` is used. not very clear just by looking at the code.

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:129]: you can probably make this whole lot simpler and remove bunch of code by using `generateName` for your Kubernetes jobs:

```
apiVersion: batch/v1
kind: Job
metadata:
  generateName: kube-job-
  labels:
    jobgroup: jobexample
```

more details here: https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:48]: can we return `taskId` as part of the error message too?

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:51]: same. lets have `taskId` in the error message.

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:59]:
```suggestion
                    log.warn("task with id $taskId not found. metadata: ${oldState.metadata} - verification: $verification")
```

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:168]: maybe can be removed with `generateName`?

File: k8s-plugin-keel/src/test/kotlin/com/amazon/spinnaker/keel/tests/K8sJobEvaluatorTest.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/test/kotlin/com/amazon/spinnaker/keel/tests/K8sJobEvaluatorTest.kt:280]: probably not necessary with `generateName`

File: k8s-plugin-keel/src/test/kotlin/com/amazon/spinnaker/keel/tests/testUtils.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/test/kotlin/com/amazon/spinnaker/keel/tests/testUtils.kt:76]: lets capture `k8s/helm@v1` and `k8s/kustomize@v1` also in constants.

File: k8s-plugin-keel/src/test/kotlin/com/amazon/spinnaker/keel/tests/testUtils.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/test/kotlin/com/amazon/spinnaker/keel/tests/testUtils.kt:118]: can we get rid of it, now?

**Review by nimakaviani — CHANGES_REQUESTED** [review]
@nabuskey can you verify my comments on the use of `generateName` before we have it merged? thanks! :)

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nabuskey) [reply k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:48]: yep

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nabuskey) [reply k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:129]: I thought about it but we don't have the job name until it's submitted and would be difficult to track which kubernetes job belongs to which verifyWith job. The metadata generated as a result of `start()` stores this information and can be used later for troubleshooting purposes.

Comment (nimakaviani) [reply k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:129]: hmm interesting. Do we have the job name from when we get the task response back from Orca? if so, we might be able to correlate.

Also how important is it for us to be able to track individual job runs back to their executions. The prefix will identify the `verifyWith` task uniquely. Do we care about differentiatings across multiple runs of the same `verifyWith`? can we make better use of timestamps etc?

Comment (nabuskey) [reply k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:129]: task ID and task name are already stored when we submit the job. We can correlate verifyWith with orca task but that's about it. The prefix probably should not be relied on to identify tasks uniquely. Users will copy/paste configurations. 
timestamp could work but it's possible for them to collide when executed at the same time. 
Same verifyWith will not run at the same time but multiple verifyWith tasks per application per environment can ran at the same time. 
If we don't care about tracing verityWith execution to k8s job, we can use `generateName`. Maybe we can just stick some diagnostic information into Jobs' metadata like execution time, env name, and artifact version. 
Thinking about it, this might be a better approach.

Comment (nimakaviani) [reply k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:129]: yeah I like where you are going with adding metadata to k8s jobs to identify tasks.

File: k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt
Comment (nimakaviani) [inline k8s-plugin-keel/src/main/kotlin/com/amazon/spinnaker/keel/k8s/verificationEvaluator/K8sJobEvaluator.kt:133]: rather than removing `NAME`, we could also append it as a suffix to the generated name, maybe? just throwing in some thoughts here. your call.

**Review by nimakaviani — APPROVED** [review]
LGTM

**Pull request merged**
